feat (Homepage): add BenefitsSection

yaroslavgorshkov opened the pull request.

yaroslavgorshkov committed:
feat (Homepage): add BenefitsSection

maxsv01 commented on src/app/components/BenefitsSection/benefitsSectionMockData.tsx:
```diff
@@ -0,0 +1,59 @@
+        descriptionTop: "We offer flexible payment",
+        descriptionBottom: "options, to make easier.",
+        svgBackgroundColor: "bg-purple",
+        svg:
```
1. почему svg идет в данных? разве svg будет присылать бекенд?
2. почему svg записан напрямую, а не в отдельном файле? почему svg не сделан как реакт компонент?

maxsv01 commented on src/app/components/BenefitsSection/benefitsSectionMockData.tsx:
```diff
@@ -0,0 +1,59 @@
+import { ReactNode } from "react";
+
+type BenefitsSectionContentItem = {
```
BenefitsSectionContentItemProps - здесь же идентичные типы, почему ты не переиспользуешь уже существующие типы?

maxsv01 replied:
так расширить, дополнить, но точно не писать их заново. Всегда стараемся переиспользовать максимально код и не повторяться.
Изучи, что такое DRY подход(Dont repeat yourself)

maxsv01 commented on src/app/components/BenefitsSection/benefitsSectionMockData.tsx:
```diff
@@ -0,0 +1,59 @@
+    text: string;
+    descriptionTop: string;
+    descriptionBottom: string;
+    svgBackgroundColor: string;
```
не тайп сейф, могу передать любой стринг

maxsv01 commented on src/app/components/BenefitsSection/benefitsSectionMockData.tsx:
```diff
@@ -0,0 +1,59 @@
+    descriptionTop: string;
+    descriptionBottom: string;
+    svgBackgroundColor: string;
+    svg: ReactNode;
```
плис, изучи более детально этот тип

```tsx
    type ReactNode =
        | ReactElement
        | string
        | number
        | bigint
        | Iterable<ReactNode>
        | ReactPortal
        | boolean
        | null
        | undefined
        | DO_NOT_USE_OR_YOU_WILL_BE_FIRED_EXPERIMENTAL_REACT_NODES[
            keyof DO_NOT_USE_OR_YOU_WILL_BE_FIRED_EXPERIMENTAL_REACT_NODES
        ]
        | Promise<AwaitedReactNode>;

```

yaroslavgorshkov committed:
fix (BenefitsSection): change BenefitsSectionContentItemProps, icon logic in benefitsSectionMockData

maxsv01 commented on src/app/components/BenefitsSection/BenefitsSectionContent/BenefitsSectionContentItem/BenefitsSectionContentItem.tsx:
```diff
@@ -0,0 +1,29 @@
+}: BenefitsSectionContentItemProps) => {
+    return (
+        <div className="flex flex-col gap-[32px] items-center w-full">
+            <div className={`flex justify-center items-center rounded-[24px] p-[20px] ${svgBackgroundColor}`} >
```
2. Всегда используем значения из готовой темы, если значения не подходят, то меняем и кастомизируем тему, но не пишем кастом знания каждый раз. Если ты захочешь подвинуть шаг в теме, то это займет пару секунд, а сейчас у тебя хард код значения в каждом компоненте. Тем более ты нарущаешь подход самого tailwind. Tailwind использует rem-ы, ты же везде хардкодно перезаписываешь на px. Это в корне не правильно.
Если уже и перезаписывать, то сделать хелпер, который тебе будет переводить px в rem и назвать его например pxToRem и использовать его только в самых необходимых местах. Сейчас ты почти не используешь тему.

maxsv01 replied:
То есть мне подогнать значения как я это сделал с кнопками? - Да, совершенно верно.
Постарайся максимально использовать тему и существующие значения. Округляй в большую или меншую сторону. Твоей целью сейчас не является точная пиксельная подгонка, поэтому смело бери приблизетельные величины.

то есть разделить отступы, педдинги, мерджины на категории (отступы секции, элемента и тд, гепы, мерджины)? - В идеале да, как и ты и описал

maxsv01 commented on src/app/components/BenefitsSection/BenefitsSectionContent/BenefitsSectionContentItem/BenefitsSectionContentItem.tsx:
```diff
@@ -0,0 +1,36 @@
+        <div className="flex flex-col gap-[32px] items-center w-full">
+            <div className={benefitsSectionContentItemIconCombinedClassName} >
+                {iconComponent}
+            </div>
```
```tsx   
 const iconComponent = getStaticIconComponent(iconName);
    const iconBackgroundColorClass = getBenefitsSectionContentItemIconBackgroundColor(iconBackgroundColor);
    const benefitsSectionContentItemIconCombinedClassName = twMerge("flex justify-center items-center rounded-[24px] p-[20px]", iconBackgroundColorClass);
    
         <div className={benefitsSectionContentItemIconCombinedClassName} >
                {iconComponent}
            </div>
```

можно вынести в отдельную компоненту. iconComponent, iconBackgroundColorClass, benefitsSectionContentItemIconCombinedClassName нужны только для этого дива, можно изолировать логику

maxsv01 commented on src/app/components/Button/Button.tsx:
```diff
@@ -21,7 +21,7 @@ export const Button = ({
     const borderRadiusClass = getBorderRadius(borderRadius);
 
-    const endIconComponent = endIcon ? getIconComponent(endIcon.iconType, endIcon.iconFillColor, endIcon.iconSize) : undefined;
+    const endIconComponent = endIcon ? getDynamicIconComponent(endIcon.iconType, endIcon.iconFillColor, endIcon.iconSize) : undefined;
```
что бы не повторяться, то можно через деструктурицазию вытняуть отдельно 

```tsx
const {iconType, iconFillColor, iconSize} =  endIcon

maxsv01 commented on src/app/components/icons.tsx:
```diff
@@ -264,4 +264,41 @@ export const FacebookIcon = ({
+    )
+}
+
+export const BenefitsSectionCustomerSupportIcon = () => {
```
все супер, только плис каждая иконка в отдельном, своем, файле....спустя пару недель и еще несколько иконок этот файл перестанет быть легким для поддержания и ориентирования в нем

yaroslavgorshkov replied:
Исправил (разнес иконки по файлам)

yaroslavgorshkov committed:
fix (BenefitsSection): distributed icons across files, moved the BenefitsSectionContentItemIcon logic into a separate component

yaroslavgorshkov committed:
fix (BenefitsSectionContentItem):
change prop type name from BenefitsSectionContentItemIconType to BenefitsSectionContentItemIconName
